[feat] #87-장소 좋아요 기능 구현

dong-yxxn opened the pull request:
### ✨ 작업 내용
- LikedPlace Entity, Repo 생성
- Controller에 좋아요 request 받는 mapping 생성
- Service에 좋아요된 장소 저장하는 로직 구현
- PlaceInfo, PlaceDetailInfo의 좋아요에 정보담는 관련 로직 생성
- 관련된 테스트코드 생성
---

### ✨ 참고 사항

---

### ⏰ 현재 버그

---

### ✏ Git Close

dong-yxxn committed:
Merge branch 'feat/#7-place' into feat/#87-liked-place

dong-yxxn committed:
[feat]: LikedPlace Entity와 Repository를 생성했어요!

- LikedPlace를 구현하기위해 관련 Entity와 Repository를 구현헀어요

관련 이슈: #87

dong-yxxn committed:
[feat] 장소 좋아요 기능 구현

- LikedPlace 엔티티와 repo를 구현했어요
- PlaceService: 장소에 좋아요하기 기능을 구현했어요
- 관련된 테스트 코드를 작성했어요

관련 이슈: #87

dong-yxxn committed:
[feat] PlaceInfo와 PlaceDetailInfo에 좋아요 여부를 반영했어요

- PlaceInfo와 PlaceDetailInfo의 likes필드에 좋아요 여부를 반영했어요
- 관련된 테스트 코드를 수정, 추가했어요

관련 이슈: #87

dong-yxxn committed:
conflict merge

dong-yxxn committed:
Merge remote-tracking branch 'origin/weekly/10' into feat/#87-liked-place

wndlthsk commented on src/main/java/team7/inplace/LikedPlace/persistence/LikedPlaceRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface LikedPlaceRepository extends JpaRepository<LikedPlace, Long> {
+
+    public Optional<LikedPlace> findByUserIdAndPlaceId(Long userId, Long placeId);
```
public 생략해도 될듯합니다

wndlthsk commented on src/main/java/team7/inplace/place/application/PlaceService.java:
```diff
@@ -146,4 +160,42 @@ public List<Long> createPlaces(List<Create> placeCommands) {
+
+        LikedPlace likedPlace = likedPlaceRepository.findByUserIdAndPlaceId(user.getId(),
+                place.getId())
+            .orElseGet(() -> new LikedPlace(user, place));
```
새로 생성된 경우에는 save 필요하지 않나요??

dong-yxxn replied:
다른 메서드에서 save하던걸 같은 메서드에서 하도록 변경했어요!

sanghee0820 commented on src/main/java/team7/inplace/LikedPlace/domain/LikedPlace.java:
```diff
@@ -0,0 +1,46 @@
+@RequiredArgsConstructor
+@NoArgsConstructor(access = PROTECTED)
+@Entity
+public class LikedPlace {
```
이게 좀 더 직관적인 것 같네요! LikedInfluencer로 수정해야겠어요

sanghee0820 commented on src/main/java/team7/inplace/global/exception/code/LikedPlaceErrorCode.java:
```diff
@@ -0,0 +1,30 @@
+@AllArgsConstructor
+@Getter
+public enum LikedPlaceErrorCode implements ErrorCode {
+    NOT_FOUND(HttpStatus.NOT_FOUND, "L001", "Can't find such liked place info");
```
한글로 적어주세요! 코파일럿이 짜주더라도 한번씩은 확인해주세요~

sanghee0820 commented on src/main/java/team7/inplace/place/application/PlaceService.java:
```diff
@@ -146,4 +160,42 @@ public List<Long> createPlaces(List<Create> placeCommands) {
+
+    private LikedPlace findOrCreateLikedPlace(Long placeId, boolean likes) {
+
+        String username = getUsername().orElseThrow(
+            () -> InplaceException.of(AuthorizationErrorCode.TOKEN_IS_EMPTY));
+
+        User user = userRepository.findByUsername(username)
+            .orElseThrow(() -> InplaceException.of(UserErroCode.NOT_FOUND));
+
+        Place place = placeRepository.findById(placeId)
+            .orElseThrow(() -> InplaceException.of(PlaceErrorCode.NOT_FOUND));
+
+        LikedPlace likedPlace = likedPlaceRepository.findByUserIdAndPlaceId(user.getId(),
+                place.getId())
+            .orElseGet(() -> new LikedPlace(user, place));
```
이 로직에서만 DB접근이 4번 이루어지고 있네요~
줄일 수 있는 방법이 없을까요?

sanghee0820 replied:
저라면, 토큰에 있는 userId와 placeId를 가지고 LikePlace DB에 존재하는지 조회하고 없다면 저장할 것 같아요~

dong-yxxn replied:
조회후, 없으면 생성하고 저장하는걸로 변경했어요

sanghee0820 commented on src/main/java/team7/inplace/place/application/PlaceService.java:
```diff
@@ -146,4 +160,42 @@ public List<Long> createPlaces(List<Create> placeCommands) {
+        String username = getUsername().orElseThrow(
+            () -> InplaceException.of(AuthorizationErrorCode.TOKEN_IS_EMPTY));
+
+        User user = userRepository.findByUsername(username)
+            .orElseThrow(() -> InplaceException.of(UserErroCode.NOT_FOUND));
```
토큰이 있는 유저인데 굳이 한번더 DB에 접근해서 사용해야할까요?

sanghee0820 reviewed (changes requested):
개인적으로는, 좋아요한 인플루언서에 대한 로직은 따로 Controller와 Service를 나누는 것도 좋아보여요. 일단 수정 요청 드린 부분 수정 부탁드립니다

dong-yxxn committed:
[refactor] 불필요한 public 접근제한자 제거

- LikedPlaceRepo에서 public 접근제어자 제거

관련 이슈: #87

dong-yxxn committed:
[refactor]: likedPlace save 메서드 위치변경으로 응집성 향상

- findOrCreateLikedPlace 호출 후, 외부에서 save하던걸 내부로 옮겨 응집성을 높였습니다.

관련 이슈: #87

dong-yxxn committed:
[refactor]: error코드 메시지를 한글로 변경함

- Place, LikedPlace ErrorCode의 메시지를 한글로 변경했습니다.

관련 이슈: #87

dong-yxxn committed:
[refactor]: userId 사용으로 db 접근 최소화

- PlaceService: findOrCreateLikedPlace에서 기존의 방법 대신, userId와 placeId로 조회 후 존재하지 않으면 placeRepo와 userRepo에 접근하여 likedPlace를 생성하도록 했습니다.
- 관련된 테스트 코드를 수정했습니다.

관련 이슈: #87

dong-yxxn committed:
Merge branch 'weekly/10' into feat/#87-liked-place

dong-yxxn committed:
[fix] 충돌 및 오타해결

wndlthsk reviewed (approved):
수고하셨어요!

sanghee0820 reviewed (approved):
고생하셨어요
